### PR TITLE
Add Content Length Decider

### PR DESCRIPTION
**Pull request by iustin24**
Added Content Length Decider

### CODE REVIEW

File: src/deciders/content_length.rs [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+cfg_if! {
+    if #[cfg(docsrs)] {
+        // just bringing in types for easier intra-doc linking during doc build
+        use crate::fuzzers::Fuzzer;
+        use crate::build_deciders;
+        use crate::deciders::Deciders;
+    }
+}
+
```
Comment (epi052) [inline src/deciders/content_length.rs:20]:
```
/// Decide upon an [`Action`] based on a [`ResponseObserver`]'s status code
///
/// # Examples
///
/// While the example below works, the normal use-case for this struct is to pass
/// it, and any other [`Deciders`] to the [`build_deciders`] macro, and pass
/// the result of that call to your chosen [`Fuzzer`] implementation.
///
/// ```
/// # use http::response;
/// # use feroxfuzz::responses::{Response, AsyncResponse};
/// # use feroxfuzz::requests::RequestId;
/// # use feroxfuzz::error::FeroxFuzzError;
/// # use tokio_test;
/// # use std::time::Duration;
/// # use feroxfuzz::corpora::Wordlist;
/// # use feroxfuzz::state::SharedState;
/// # use feroxfuzz::actions::Action;
/// # use feroxfuzz::deciders::ContentLengthDecider;
/// # use feroxfuzz::deciders::Decider;
/// # use feroxfuzz::observers::Observers;
/// # use feroxfuzz::observers::Observer;
/// # use feroxfuzz::observers::ResponseObserver;
/// # use feroxfuzz::build_observers;
/// # use feroxfuzz::deciders::{DeciderHooks, LogicOperation};
/// # use feroxfuzz::MatchName;
/// # fn main() -> Result<(), FeroxFuzzError> {
/// # tokio_test::block_on(async {
/// // for testing; normally a Response comes as a result of a sent request
/// let reqwest_response = http::response::Builder::new().body("0123456789").unwrap();
/// let id = RequestId::new(0);
/// let elapsed = Duration::from_secs(1);
/// let response = AsyncResponse::try_from_reqwest_response(id, reqwest_response.into(), elapsed).await?;
///
/// // also not relevant to the current example, but needed to make the call to .post_send_hook
/// let corpus = Wordlist::with_words(["a", "b", "c"]).name("chars").build();
/// let mut state = SharedState::with_corpus(corpus);
///
///
/// // a ResponseObserver should be already created at some earlier point
/// let response_observer: ResponseObserver<_> = response.into();
/// let observers = build_observers!(response_observer);
///
/// // create a ContentLengthDecider with a size of 10 and a closure
/// // that will provide the 'how' of the decision making process
/// let mut decider = ContentLengthDecider::new(10, |length, observed, _state| {
///     if length == observed {
///         Action::Keep
///     } else {
///         Action::Discard
///     }
/// });
///
/// // finally, make the call to .post_send_hook and receive the answer in the form of
/// // an Action to take
/// let action = decider.post_send_hook(&mut state, &observers, Some(Action::Keep), LogicOperation::Or);
///
/// assert_eq!(action, Some(Action::Keep));
/// # Result::<(), FeroxFuzzError>::Ok(())
/// # })
/// # }
/// ```
```

please add this docstring, and confirm that it passes by running `cargo test --all-features --doc`

File: src/deciders/content_length.rs [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+{
+    comparator: F,
+    content_length: usize,
+}
+
+impl<F> ContentLengthDecider<F>
+    where
+        F: Fn(usize, usize, &SharedState) -> Action,
+{
+    pub const fn new(content_length: usize, comparator: F) -> Self {
```
Comment (epi052) [inline src/deciders/content_length.rs:36]:
```
/// create a new `ContentLengthDecider` that calls `comparator` in its
/// `post_send_hook` method
```

please add the docstring

File: src/deciders/content_length.rs [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+            // get the observed content length
+            let observed_content_length = observer.content_length();
+
+            // call the comparator to arrive at a decided action
+            return Some((self.comparator)(self.content_length, observed_content_length, state));
+        }
+
+        None
+    }
+}
```
Comment (epi052) [inline src/deciders/content_length.rs:71]:
```
impl<F> AsAny for ContentLengthDecider<F>
where
    F: Fn(usize, usize, &SharedState) -> Action + 'static,
{
    fn as_any(&self) -> &dyn Any {
        self
    }
}

impl<F> Named for ContentLengthDecider<F>
where
    F: Fn(usize, usize, &SharedState) -> Action + 'static,
{
    fn name(&self) -> &'static str {
        "ContentLengthDecider"
    }
}
```

please add the two traits above, I'm adding dynamic dispatch support in another branch and these will need to be there at some point. It's also the same reason for the extra constraints on the closure i added above

File: src/deciders/content_length.rs [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        R: Response,
+        F: Fn(usize, usize, &SharedState) -> Action,
+{
+}
+
+impl<O, R, F> Decider<O, R> for ContentLengthDecider<F>
+    where
+        O: Observers<R>,
+        R: Response,
+        F: Fn(usize, usize, &SharedState) -> Action,
```
Comment (epi052) [inline src/deciders/content_length.rs:56]: `F: Fn(usize, usize, &SharedState) -> Action + Sync + Send + Clone + 'static`

File: src/deciders/content_length.rs [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+            content_length,
+        }
+    }
+}
+
+impl<O, R, F> DeciderHooks<O, R> for ContentLengthDecider<F>
+    where
+        O: Observers<R>,
+        R: Response,
+        F: Fn(usize, usize, &SharedState) -> Action,
```
Comment (epi052) [inline src/deciders/content_length.rs:48]: `F: Fn(usize, usize, &SharedState) -> Action + Sync + Send + Clone + 'static`

File: src/deciders/content_length.rs [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        use crate::deciders::Deciders;
+    }
+}
+
+#[derive(Clone, PartialEq, Eq, Hash, PartialOrd, Ord, Debug)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[allow(clippy::derive_partial_eq_without_eq)] // known false-positive introduced in 1.63.0
+pub struct ContentLengthDecider<F>
+    where
+        F: Fn(usize, usize, &SharedState) -> Action,
```
Comment (epi052) [inline src/deciders/content_length.rs:26]: `F: Fn(usize, usize, &SharedState) -> Action + Sync + Send + Clone + 'static`

**Review by epi052 — CHANGES_REQUESTED** [review]
howdy, and thanks for taking the time to submit the (project's first) PR! :tada: 

I asked for a few changes, but provided the code, so it shouldn't be too terrible. 

Once you're done, could you please run `cargo fmt`? Or, if you have `cargo-make` installed, just run `cargo make check`. It'll run clippy/fmt/tests/docs.

Thanks again!

File: src/deciders/content_length.rs (hunk 6 above)
Comment (epi052) [reply src/deciders/content_length.rs:26]: I don't think adding the extra constraints will break anything, but they may. If they do break other things, don't worry about it,  I'll get them in my branch.

**Comment by iustin24** [comment]
Hi Epi052,

I added the changes and ran `cargo make check` with all tests passing.
I didn't add the two traits required for dynamic dispatch support since they were causing compile errors.

Let me know if I missed anything!

**Comment by epi052** [comment]
lgtm, thanks again! 

@all-contributors add @iustin24 for code

**Pull request merged**